### PR TITLE
Fix the signature of CRYPT_EX_dup

### PR DESCRIPTION
**Pull request by bernd-edlinger**
This fixes a strict aliasing issue in ui_dup_method_data.

And fix the error handling in CRYPTO_dup_ex_data.

I have split the error handling fix in an extra commit,
as the function signature can probably only be changed
in the master branch, while the error handling should
be fixed in 1.1.0 as well.  I will need another PR for 1.0.2.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/CRYPTO_get_ex_new_index.pod [hunk 1]
```diff
@@ -130,9 +130,9 @@ the same callback handles different types of exdata.
 dup_func() is called when a structure is being copied.  This is only done
 for B<SSL> and B<SSL_SESSION> objects.  The B<to> and B<from> parameters
 are pointers to the destination and source B<CRYPTO_EX_DATA> structures,
-respectively.  The B<srcp> parameter is a pointer to the source exdata.
-When the dup_func() returns, the value in B<srcp> is copied to the
-destination ex_data.  If the pointer contained in B<srcp> is not modified
+respectively.  The B<*srcp> parameter is a pointer to the source exdata.
```
Comment (bernd-edlinger) [inline doc/man3/CRYPTO_get_ex_new_index.pod]: aehm, this should probably refer to from_d, as the parameter is called in the prototype..

**Comment by richsalz** [comment]
sigh.  and if we don't fix the API, what's the risk?  we really hate changing API's unless it's a serious bug being fixed.

File: include/openssl/crypto.h [hunk 2]
```diff
@@ -175,7 +175,7 @@ typedef void CRYPTO_EX_new (void *parent, void *ptr, CRYPTO_EX_DATA *ad,
 typedef void CRYPTO_EX_free (void *parent, void *ptr, CRYPTO_EX_DATA *ad,
                              int idx, long argl, void *argp);
 typedef int CRYPTO_EX_dup (CRYPTO_EX_DATA *to, const CRYPTO_EX_DATA *from,
-                           void *srcp, int idx, long argl, void *argp);
+                           void **from_d, int idx, long argl, void *argp);
```
Comment (levitte) [inline include/openssl/crypto.h]: Sadly, this can only be changed in a major version change. The next one is 1.2.0

**Comment by bernd-edlinger** [comment]
In theory the compiler can assume that *pptr in ui_dup_method_data
is not an alias of ptr in CRYPTO_dup_ex_data, because char * and void *
can not alias, but I would not fix it in 1.1.0, only in master.

**Comment by richsalz** [comment]
As @levitte points out, it is an API change and therefore can't go in master now since that's going to be 1.1.1  So this will have to wait, likely until 2018.

File: include/openssl/crypto.h (hunk 2 above)
Comment (bernd-edlinger) [inline include/openssl/crypto.h]: so you mean that can not be fixed in master ?

**Comment by richsalz** [comment]
Not now, no. WE committed to our next release being 1.1.1 which is ABI compatible.

**Comment by bernd-edlinger** [comment]
Ok, I see, actually it is ok for the strict aliasing to cast a void **
to a void* as long as you cast it back to a void **

File: include/openssl/crypto.h (hunk 2 above)
Comment (levitte) [inline include/openssl/crypto.h]: This change can be kept in a separate PR that eats marked with the 1.2.0 milestone.

**Comment by levitte** [comment]
Are you expecting all implementations to start casting `srcp` to `void **` all of a sudden?

**Comment by bernd-edlinger** [comment]
aehm, maybe they read the documentation :-)

**Comment by bernd-edlinger** [comment]
I could write something like:
The B<from_d> parameter needs to be cast to a B<void **pptr>
as the API has currently the wrong signature; that will be changed in a 
future version.  The B<*pptr> is a pointer to the source exdata.
When the dup_func() returns, the value in B<*pptr> is copied to the
destination ex_data.

**Comment by bernd-edlinger** [comment]
rebased, and waiting for 1.2.0 now...

**Comment by bernd-edlinger** [comment]
> So this will have to wait, likely until 2018.

By the way it is 2020 now.

File: doc/man3/CRYPTO_get_ex_new_index.pod [hunk 3]
```diff
@@ -140,10 +140,8 @@ dup_func() is called when a structure is being copied.  This is only done
 for B<SSL>, B<SSL_SESSION>, B<EC_KEY> objects and B<BIO> chains via
 BIO_dup_chain().  The B<to> and B<from> parameters
 are pointers to the destination and source B<CRYPTO_EX_DATA> structures,
-respectively.  The B<from_d> parameter needs to be cast to a B<void **pptr>
-as the API has currently the wrong signature; that will be changed in a
-future version.  The B<*pptr> is a pointer to the source exdata.
-When the dup_func() returns, the value in B<*pptr> is copied to the
+respectively.  The B<*from_d> parameter is a pointer to the source exdata.
+When the dup_func() returns, the value in B<*from_d> is copied to the
```
Comment (levitte) [inline doc/man3/CRYPTO_get_ex_new_index.pod:144]: Please add a history section that mentions this change.  Mark it for 3.0.0

**Comment by levitte** [comment]
This is ok for master.  Considering that the current documentation says that the passed argument should really be a `void **`, and that the API will be fixed in the future, 3.0 is as good a future as any.

**Comment by levitte** [comment]
When history section has been added, this should be approvable

File: doc/man3/CRYPTO_get_ex_new_index.pod (hunk 3 above)
Comment (bernd-edlinger) [reply doc/man3/CRYPTO_get_ex_new_index.pod:144]: does the history section look okay?

Comment (levitte) [reply doc/man3/CRYPTO_get_ex_new_index.pod:144]: Yup

**Comment by levitte** [comment]
Just to be a bit extra explicit, I approve this for master only.

**Comment by bernd-edlinger** [comment]
Merged to master as 712e8de

**Comment by kaduk** [comment]
Of course, `void**` does not have the `void*` nature of "compatible with any pointer type of same constedness/etc", so implementations will have to cast on assignment to their local type to get a warning-free build.  But that's not particularly hard to do.
